pyccel/complexity/basic.py

```
@@ -25,16 +38,230 @@ def __init__(self, filename_or_text):
+    def mode(self):
+        return self._mode
+
+    @property
+    def visual(self):
+        return self._visual
+
+    def cost(self, visual=True, mode=None, simplify=True, bigo=None):
+        """
+        Computes the complexity of the given code.
+
+        verbose: bool
+            talk more
+
+        mode: string
+            possible values are (None, simple)
+
+        simplify: bool
+            if true, calls sympy simplify function
+
+        bigo: list
+            a list containing variable names (str) for which we want to take the
+            highest degree (expansion serie)
+        """
+        # ...
+        self._visual = visual
+        self._mode = mode
+        # ...
+
+        # ...
+        costs = OrderedDict()
+
+        # ... first we treat declared functions
+        if self.functions:
+            for fname, d in self.functions.items():
+                expr =  self._cost(d)
+
+                if simplify:
+                    expr = sp_simplify(expr)
+
+                if bigo:
+                    for n in bigo:
+                        expr = LT(expr, Symbol(n))
+
+                costs[fname] = expr
+
+        self._costs.update(costs)
+        # ...
+
+        # ... then we compute the complexity for the main program
+        expr = self._cost(self.ast)
+
+        if simplify:
+            expr = sp_simplify(expr)
+
+        if bigo:
+            for n in bigo:
+                expr = LT(expr, Symbol(n))
+        # ...
+
+        return expr
+
+    def _cost(self, expr, **settings):
+        if expr is None:
+            return 0
+
+#        print('>>> ', expr, type(expr))
+
+        classes = type(expr).__mro__
+        for cls in classes:
+            method = '_cost_' + cls.__name__
+            if hasattr(self, method):
+                obj = getattr(self, method)(expr, **settings)
+                return obj
+            else:
+                raise NotImplementedError('{} not available for {}'.format(method, type(expr)))
```

Shouldn't this `raise` be after the for block?

---

I think you have misunderstood something. With the behaviour that you have implemented the for loop is useless. I assume the behaviour is supposed to match what is done in the semantic and syntactic parser and the code printers. The way it works there is as follows:
Suppose we want to print `NumpyTanh` first we collect the inheritance tree of `NumpyTanh` which gives us:
`('NumpyTanh', 'NumpyUfuncUnary', 'NumpyUfuncBase', 'PyccelInternalFunction', 'PyccelAstNode', 'Basic')`
This means that the cost functions which are acceptable for visiting this object are:
- `_cost_NumpyTanh`
- `_cost_NumpyUfuncUnary`
- `_cost_NumpyUfuncBase`
- `_cost_PyccelInternalFunction`
- `_cost_PyccelAstNode`
- `'_cost_Basic`

We run through these possible functions choosing the one which is the most specialised and only if none of these methods exist do we need to raise an error

---

Thank you